src/librustc_mir/build/matches/test.rs

```diff
@@ -205,7 +205,9 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 if let Some(otherwise_block) = otherwise_block {
                     targets.push(otherwise_block);
                 } else {
-                    values.pop();
+                    let unreachable_block = self.cfg.start_new_block();
```

Comment:
Presuming we do decide to keep this approach, this particular patch doesn't feel like the best:

- It relies on the fact that `start_new_block` is returning a "unreachable" terminator by default. That feels like an implementation detail we may want to change.
- It creates an extra block per match, when I suppose we only need one for the entire function, right?

I'd rather have something like `self.cfg.unreachable_block()` that lazilly (or eagerly...) creates such a block.

Reply:
As I understand it, the blocks are not actually populated at all in this step. All of them are `start_new_block`. The code that then follows on from this goes through each case and fills them in. It already handles the "last case is unreachable case" because it already has to handle

```rust
let x: u8;
match x {
    0...255 => ...,
    _ => ...
}
```

Reply:
> It creates an extra block per match, when I suppose we only need one for the entire function, right?

Not sure what you mean here? Surely we could be matching different things within the same function

Reply:
So, I might have been remembering wrong in terms of how things work, I'll have to check -- but I don't think I am wrong that we only need one unreachable block per function. That is, why would we need more than one? All matches can share the same one, I should think.

Reply:
Ah I see what you mean now - I will investigate how many blocks get generated for a function with many matches. I will note that if multiple are generated for the same match statement, they will be removed. See matches/mod.rs:94:

```
        // this will generate code to test discriminant_lvalue and
        // branch to the appropriate arm block
        let otherwise = self.match_candidates(span, &mut arm_blocks, candidates, block);

        if !otherwise.is_empty() {
            // All matches are exhaustive. However, because some matches
            // only have exponentially-large exhaustive decision trees, we
            // sometimes generate an inexhaustive decision tree.
            //
            // In that case, the inexhaustive tips of the decision tree
            // can't be reached - terminate them with an `unreachable`.
            let source_info = self.source_info(span);

            let mut otherwise = otherwise;
            otherwise.sort();
            otherwise.dedup(); // variant switches can introduce duplicate target blocks
            for block in otherwise {
                self.cfg.terminate(block, source_info, TerminatorKind::Unreachable);
            }
}
```

Reply:
Hm have run into a snag with this. A `Terminator` has a `TerminatorKind` as well as a `SourceInfo`. Since `SourceInfo` necessarily differs between different matches, it is not possible to dedup the blocks, unless we put the wrong `SourceInfo` on some (unreachable) basic blocks.

Reply:
So we do something similar for the `cached_return_block`. The idea is that everybody references this block, which has no terminator as of yet, and then we assign the terminator with the source info of the function end:

https://github.com/rust-lang/rust/blob/master/src/librustc_mir/build/mod.rs#L400-L401

We might do something like that here.